--- ./../build/libdeps/logging-log4cxx-0_10_0/src/main/include/log4cxx/private/Makefile.am	(revision 1143468)
+++ ./../build/libdeps/logging-log4cxx-0_10_0/src/main/include/log4cxx/private/Makefile.am	(working copy)
@@ -14,7 +14,7 @@
 # limitations under the License.
 #
 privateincdir = $(includedir)/log4cxx/private
-privateinc_HEADERS= $(top_builddir)/src/main/include/log4cxx/private/*.h log4cxx_private.h
+privateinc_HEADERS= $(top_builddir)/src/main/include/log4cxx/private/*.h
 DISTCLEANFILES = log4cxx_private.h
 EXTRA_DIST = log4cxx_private.hw
 
