--- ./../build/libdeps/logging-log4cxx-0_10_0/src/main/include/log4cxx/Makefile.am (revision 1143468)
+++ ./../build/libdeps/logging-log4cxx-0_10_0/src/main/include/log4cxx/Makefile.am (working copy)
@@ -15,7 +15,7 @@
 #
 SUBDIRS = helpers net nt rolling spi varia xml config db private pattern filter
 log4cxxincdir = $(includedir)/log4cxx
-log4cxxinc_HEADERS= $(top_srcdir)/src/main/include/log4cxx/*.h log4cxx.h
+log4cxxinc_HEADERS= $(top_srcdir)/src/main/include/log4cxx/*.h
 DISTCLEANFILES = log4cxx.h
 EXTRA_DIST = log4cxx.hw
